db/snapshot/backup_task.cc

```diff
@@ -103,52 +105,115 @@ future<> backup_task_impl::upload_component(shared_ptr<s3::client>& client, sstr
+            auto shard_id = this_shard_id();
+            gate uploads;
+            std::exception_ptr ex;
+            try {
+                auto cln = _snap_ctl.storage_manager().container().local().get_endpoint_client(_endpoint);
+
+                // Fetching files sequentially within each shard allows less loaded shards to take on more work, compensating for those that are lagging behind.
+                auto backup_file_idx = [&counter, backup_size = backup_files.size()] -> std::optional<size_t> {
+                    auto ret_val = counter++;
+                    if (ret_val < backup_size) {
+                        return ret_val;
+                    }
+                    return std::nullopt;
+                };
+
+                while (auto name_idx = backup_file_idx()) {
+                    auto gh = uploads.hold();
+                    // Pre-upload break point. For testing abort in actual s3 client usage.
+                    co_await utils::get_local_injector().inject("backup_task_pre_upload", utils::wait_for_message(std::chrono::minutes(2)));
+
+                    // It is safe to ignore this `future` here since the `cln` is a shared pointer and the file name are both passed by value, and the abort
+                    // source has a scope of the current function. Also, the gate keeps the fiber being executed before the `as` is destroyed by the scope.
+                    std::ignore = upload_component(cln, aborter, _progress_per_shard[shard_id], std::get<1>(backup_files[*name_idx]))
+                                      .handle_exception([&aborter](std::exception_ptr e) {
+                                          // Save the exception to abort the `while` loop and later propagate the exception to other shards.
+                                          aborter.request_abort_ex(std::move(e));
+                                      }).finally([gh = std::move(gh)] {});
+
+                    co_await coroutine::maybe_yield();
+                    co_await utils::get_local_injector().inject("backup_task_pause", utils::wait_for_message(std::chrono::minutes(2)));
+                    if (aborter.abort_requested()) {
+                        break;
+                    }
+                }
+            } catch (...) {
+                ex = std::current_exception();
+            }
+            co_await uploads.close();
+
+            if (ex || aborter.abort_requested()) {
+                // Either an exception occurred or the backup was aborted or ignored `future` from `upload_component` got an exception. Propagate the exception
+                // to other shards and rethrow for the exception to propagate to task's abort source.
+                if (!ex) {
+                    ex = aborter.abort_requested_exception_ptr();
+                }
+                co_await sharded_aborter.invoke_on_all([&ex](abort_source& abort_service) -> future<> {
```

Comment:
This causes smp::count-squared storm of abort requsts -- once _any_ (including `task._as` one) abort source is aborted, all shards get here and abort on all shards again